ruma-events: Change the structure of key.verification.start events.

poljar opened the pull request:
As promised a pr for #160 is here.

I think we might want to change the accept event to have a similar structure to this, if so we should probably make this part of this PR as well.

edit(jplatte):

* fixes #160 
* fixes #165

mujx commented:
Should we handle the `reason` field in another PR or wait for the spec?

farodin91 commented:
We can handle this field in another PR. We should open an issue after opening we can merge this.

poljar committed:
ruma-events: Change the structure of key.verification.start events.

jplatte commented on ruma-events/src/key/verification/start.rs:
```diff
@@ -281,14 +317,18 @@ mod tests {
-                short_authentication_string,
-            }) if from_device == "123"
+                method: StartMethod::MSasV1(
+                    MSasV1Content {
```
Please collapse these:

```suggestion
                method: StartMethod::MSasV1(MSasV1Content {
```

jplatte commented on ruma-events/src/key/verification/start.rs:
```diff
@@ -315,21 +355,55 @@ mod tests {
-                    short_authentication_string,
-                })
+                    method: StartMethod::MSasV1(
+                        MSasV1Content {
```
Same here.

jplatte reviewed (commented):
> I think we might want to change the accept event to have a similar structure to this

Yes, updating the accept event would be nice.

poljar committed:
ruma-events: Change the structure of the accept event as well.

This allows the accept event to contain method specific content similar
to how the start event accepts it.

poljar committed:
ruma-events: Implement TryFrom to for MSasV1Content.

poljar committed:
ruma-events: Rename MSasV1ContentOptions to MSasV1ContentInit.

jplatte commented on ruma-events/src/key/verification/accept.rs:
```diff
@@ -1,11 +1,13 @@
+
 use ruma_events_macros::BasicEventContent;
 use serde::{Deserialize, Serialize};
+use serde_json::Value;
```
Please rename like everywhere else.

```suggestion
use serde_json::Value as JsonValue;
```

jplatte commented on ruma-events/src/key/verification/accept.rs:
```diff
@@ -20,33 +22,235 @@ pub type AcceptEvent = BasicEvent<AcceptEventContent>;
+                        message_authentication_code,
+                        short_authentication_string,
+                    }
+                )
```
This is not indented correctly. Unfortunately, `rustfmt` doesn't know how to format `assert_matches!` calls so it mostly leaves them alone.

jplatte reviewed (commented):
Thanks for doing the rename as well :)

EDIT: Oh, please don't do issue references into commits though, this has caused issues for us before when we imported commits from one of the smaller repositories. Please put that into the PR description.

poljar committed:
ruma-events: Add the hmac-sha256 MAC method.

poljar commented:
Since we now have a `MSasV1Content` for the accept event, we'll need a init struct as well, right? Or do we want to remove the non-exhaustive attribute?

jplatte commented on ruma-events/src/key/verification.rs:
```diff
@@ -42,6 +42,8 @@ pub enum KeyAgreementProtocol {
 pub enum MessageAuthenticationCode {
     /// The HKDF-HMAC-SHA256 MAC.
     HkdfHmacSha256,
+    /// The HMAC-SHA256 MAC.
+    HmacSha256,
```
The spec doesn't actually mention this, does it?

poljar replied:
Sure, it's a deprecated method (or perhaps it never made it into the spec, can' remember) but the events are completely useless otherwise, of course unless we resolve https://github.com/ruma/ruma/issues/171.

jplatte commented:
> Since we now have a `MSasV1Content` for the accept event, we'll need a init struct as well, right? Or do we want to remove the non-exhaustive attribute?

We generally want `#[non_exhaustive]` for things can have fields added in a future spec version in a backwards-compatible way (or there's extra invariants to ensure in construction). Which is pretty much everything. Init structs should exist when it's more readable than a regular constructor function. With the large amount of mandatory fields, this seems to be the case for `accept::MSasV1Content`.

poljar committed:
ruma-events: Add an init struct for the m.sas.v1 content of the accept event.

jplatte commented:
Please rebase to squash the fixup commit & formatting changes and remove the issue references from the commit messages.

jplatte commented:
Thanks!
